// @ractivejs/cli/src/scaffolds.js
export const component = `
<link rel="ractive" href="">

<template>

</template>

<style>

</style>

<script>

</script>
`.slice(1)

export const constructor = `
module.exports = Ractive.extend({
  data: {

  },
  components: {

  }
  template: \`

  \`,
  css: \`

  \`
})
`.slice(1)

export const adaptor = `
module.exports = {
  filter (value, keypath, ractive) {

  },
  wrap (ractive, value, keypath, prefixer) {

    return {
      get () {

      },
      set (property, value) {

      },
      reset (value) {

      },
      teardown () {

      }
    }
  }
}
`.slice(1)

export const decorator = `
module.exports  = (node, ...args) => {

  return {
    teardown () {

    },
    update (...args) {

    },
    invalidate () {

    }
  }
}
`.slice(1)

export const easing = `
module.exports = time => {

}
`.slice(1)

export const interpolator = `
module.exports = (from, to) => progress => {

}
`.slice(1)

export const event = `
module.exports = (node, fire) => {

  return {
    teardown () {

    }
  }
}
`.slice(1)

export const transition = `
module.exports = (transitionHelper, ...args) => {

}
`.slice(1)
